llvm/lib/Transforms/InstCombine/InstCombineAddSub.cpp

```diff
@@ -2807,6 +2807,44 @@ Instruction *InstCombinerImpl::visitSub(BinaryOperator &I) {
   if (Instruction *Res = foldBinOpOfSelectAndCastOfSelectCondition(I))
     return Res;
 
+  // (sub (sext (add nsw (X, Y)), sext (X))) --> (sext (Y))
+  if (match(Op1, m_SExt(m_Value(X))) &&
+      match(Op0, m_SExt(m_c_NSWAdd(m_Specific(X), m_Value(Y))))) {
```

Comment:
```suggestion
  if (match(Op1, m_SExtLike(m_Value(X))) &&
      match(Op0, m_SExtLike(m_c_NSWAdd(m_Specific(X), m_Value(Y))))) {
```
Can you add some tests with `zext nneg`?

Reply:
Will do. Thanks!